read-fonts/src/tables/postscript/stack.rs

```diff
@@ -0,0 +1,381 @@
+//! Operand stack for CFF/CFF2 parsing.
+
+use types::Fixed;
+
+use super::{BlendState, Error};
+
+/// Maximum size of the operand stack.
+///
+/// "Operators in Top DICT, Font DICTs, Private DICTs and CharStrings may be
+/// preceded by up to a maximum of 513 operands."
+///
+/// <https://learn.microsoft.com/en-us/typography/opentype/spec/cff2#table-9-top-dict-operator-entries>
+const MAX_STACK: usize = 513;
+
+/// Operand stack for DICTs and charstrings.
+///
+/// The operand stack can contain either 32-bit integers or 16.16 fixed point
+/// values. The type is known when pushing to the stack and the expected type
+/// is also known (based on the operator) when reading from the stack, so the
+/// conversion is performed on demand at read time.
+///
+/// Storing the entries as an enum would require 8 bytes each and since these
+/// objects are created on the _stack_, we reduce the required size by storing
+/// the entries in parallel arrays holding the raw 32-bit value and a flag that
+/// tracks which values are fixed point.
+pub struct Stack {
+    values: [i32; MAX_STACK],
+    value_is_fixed: [bool; MAX_STACK],
+    top: usize,
+}
+
+impl Stack {
+    pub fn new() -> Self {
+        Self {
+            values: [0; MAX_STACK],
+            value_is_fixed: [false; MAX_STACK],
+            top: 0,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.top == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.top
+    }
+
+    pub fn verify_exact_len(&self, len: usize) -> Result<(), Error> {
+        if self.top != len {
+            Err(Error::StackUnderflow)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn verify_at_least_len(&self, len: usize) -> Result<(), Error> {
+        if self.top < len {
+            Err(Error::StackUnderflow)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Returns true if the number of elements on the stack is odd.
+    ///
+    /// Used for processing some charstring operators where an odd
+    /// count represents the presence of the glyph advance width at the
+    /// bottom of the stack.
+    pub fn len_is_odd(&self) -> bool {
+        self.top & 1 != 0
+    }
+
+    pub fn clear(&mut self) {
+        self.top = 0;
+    }
+
+    /// Reverse the order of all elements on the stack.
+    ///
+    /// Some charstring operators are simpler to process on a reversed
+    /// stack.
+    pub fn reverse(&mut self) {
+        self.values[..self.top].reverse();
+        self.value_is_fixed[..self.top].reverse();
+    }
+
+    pub fn push_i32(&mut self, value: i32) -> Result<(), Error> {
+        self.push(value, false)
+    }
+
+    pub fn push_fixed(&mut self, value: Fixed) -> Result<(), Error> {
+        self.push(value.to_bits(), true)
+    }
```

Comment:
One other API possibility here is to use an enum in the API, but not in the internal storage.

This would look like,

```rust
struct Operand {
    I32(i32),
    Fixed(Fixed),
}

// in our impl

pub fn push(&mut self, value: impl Into<Operand>) -> Result<(), Error> {
    match value.into() {
        Operand::I32(val) => self.push(val, false),
        Operand::Fixed(val) => self.push(val.to_bits(), false),
    }
}
```

again, just spitballing.

Reply:
Good call! I actually have a `Number` enum in code that hasn’t been pushed yet that looks just like this. I’ll add it to this PR and use your unified push method.